### PR TITLE
Go modules: raise GitDependenciesNotReachable

### PR DESCRIPTION
**Pull request by thepwagner**
Sharing snapshot to discuss the approach:

When Dependabot encounters a Go module backed by a Git repository, especially one on `github.com`, it should raise `Dependabot::GitDependenciesNotReachable` with `dependency-urls` details describing the missing repository. This matters as the rich exception provides more detail about what went wrong and potential recovery options.

This PR accomplishes this by intercepting the error thrown from the "UpdateChecker", and detecting `github.com/.*` urls. If we detect the failed dependency was on `github.com`, we assume it's a "git dependency" and raise accordingly.

### CODE REVIEW

File: go_modules/lib/dependabot/go_modules/update_checker.rb [hunk 1]
```diff
@@ -86,6 +88,11 @@ def find_latest_resolvable_version
 
       def handle_subprocess_error(error)
         if RESOLVABILITY_ERROR_REGEXES.any? { |rgx| error.message =~ rgx }
+          if error.message.start_with?("module github.com/")
```
Comment (jurre) [inline go_modules/lib/dependabot/go_modules/update_checker.rb:91]: Should we even care about enterprise at this point? Probably YAGNI for now

File: go_modules/lib/dependabot/go_modules/update_checker.rb [hunk 2]
```diff
@@ -86,6 +88,11 @@ def find_latest_resolvable_version
...
+            raise Dependabot::GitDependenciesNotReachable, [
+              error.message.scan(%r{github.com/[^:]*}).last
```
Comment (jurre) [inline go_modules/lib/dependabot/go_modules/update_checker.rb:93]: Nit: might pull the regex out into a constant

**Review by jurre — APPROVED** [review]
This seems like a reasonable approach to me 👍

File: go_modules/lib/dependabot/go_modules/update_checker.rb (hunk 1 above)
Comment (thepwagner) [reply go_modules/lib/dependabot/go_modules/update_checker.rb:91]: AFAICT the only hook we have for that is the `api_endpoint` for octokit?: https://github.com/dependabot/dependabot-core/blob/05649b25af72d406b144965b6498167351547081/common/lib/dependabot/clients/github_with_retries.rb#L38

Getting the `Dependabot::Source` out of the `Dependabot::Dependency` that's passed here [seems complicated](https://github.com/dependabot/dependabot-core/blob/05649b25af72d406b144965b6498167351547081/common/lib/dependabot/git_commit_checker.rb#L233-L249), converting the API endpoint to the module path is another problem.

I'm also not sure about assuming GHES customers include the server URL in included module paths? 🤷 
For Enterprise-y reasons, I may want to host my server at `github.acme.com`, but namespace my modules as `acme.com/**`?

Definitely YAGNI.

Comment (fatih) [reply go_modules/lib/dependabot/go_modules/update_checker.rb:91]: This looks good for now, but I'm wondering the implications for modules hosted on GitHub but have a different module name by using a proxy.  We're probably going to found out once we start using this in production.

**Review by fatih — APPROVED** [review]
Looks good to me 👍🏼

File: go_modules/lib/dependabot/go_modules/file_parser.rb [hunk 3]
```diff
@@ -156,6 +157,26 @@ def handle_parser_error(path, stderr)
         end
       end
 
+      GITHUB_REPO_REGEX = %r{github.com/[^@]*}.freeze
+      def handle_github_unknown_revision(line)
+        repo_path = line.scan(GITHUB_REPO_REGEX).first
+        return unless repo_path
```
Comment (thepwagner) [inline go_modules/lib/dependabot/go_modules/file_parser.rb:163]:
```suggestion
```

Nit: maybe too defensive? Given the guard on https://github.com/dependabot/dependabot-core/pull/2780/files#diff-5473cd4fd4f2f0f6b13968468fc3a0109a7769f9c729e4ad36a03ac8f419a660R143 this is never true.

File: go_modules/lib/dependabot/go_modules/file_parser.rb [hunk 4]
```diff
@@ -139,8 +139,9 @@ def go_mod_content
       def handle_parser_error(path, stderr)
         case stderr
         when /go: .*: unknown revision/m
-          line = stderr.lines.grep(/unknown revision/).first
-          raise Dependabot::DependencyFileNotResolvable, line.strip
+          line = stderr.lines.grep(/unknown revision/).first.strip
+          handle_github_unknown_revision(line) if line.start_with?("go: github.com/")
```
Comment (thepwagner) [inline go_modules/lib/dependabot/go_modules/file_parser.rb:143]: Question: should this behaviour still be specific to `github.com/.*` urls?

Raising a typed exception for inaccessible module dependencies should be useful regardless of host, and the pattern seems portable.

I kept this filter for defensiveness.

Comment (jurre) [reply go_modules/lib/dependabot/go_modules/file_parser.rb:143]: In theory they could be mercurial or bazaar deps?

File: go_modules/lib/dependabot/go_modules/file_parser.rb (hunk 3 above)
Comment (jurre) [reply go_modules/lib/dependabot/go_modules/file_parser.rb:163]: If we keep `line.start_with?("go: github.com/")` seems like we could get rid of it, but I don't mind the check either

**Review by jurre — APPROVED** [review]

**Pull request merged**